onnxruntime/core/providers/cuda/cuda_execution_provider.cc

```diff
@@ -34,54 +35,73 @@ class Memcpy final : public OpKernel {
       ORT_ENFORCE(X != nullptr, "Memcpy: Input tensor is nullptr.");
       Tensor* Y = ctx->Output(0, X->Shape());
       ORT_ENFORCE(Y != nullptr, "Memcpy: Failed to allocate output tensor.");
-      return Info().GetDataTransferManager().CopyTensor(*X, *Y, Info().GetKernelDef().ExecQueueId());
-    } else if (X_type->IsSparseTensorType()) {
-      const auto* X = ctx->Input<SparseTensor>(0);
-      ORT_ENFORCE(X != nullptr, "Memcpy: Input tensor is nullptr.");
-      SparseTensor* Y = ctx->OutputSparse(0, X->DenseShape());
-      ORT_ENFORCE(Y != nullptr, "Memcpy: Failed to allocate output sparse tensor.");
-      return X->Copy(Info().GetDataTransferManager(), Info().GetKernelDef().ExecQueueId(), *Y);
-    } else if (X_type->IsTensorSequenceType()) {
-      const TensorSeq* X = ctx->Input<TensorSeq>(0);
-      ORT_ENFORCE(X != nullptr, "Memcpy: Input tensor sequence is nullptr.");
-      TensorSeq* Y = ctx->Output<TensorSeq>(0);
-      ORT_ENFORCE(Y != nullptr, "Memcpy: Failed to allocate output tensor sequence.");
-      auto X_dtype = X->DataType();
-      Y->SetType(X_dtype);
-      AllocatorPtr alloc;
-
-      // If we are copying contents to CUDA, the allocator to use
-      // to allocate the buffers of the new tensors in the sequence
-      // can be temp space allocator associated with the CUDA EP
-      if (Node().OpType() == "MemcpyFromHost") {
-        auto status = ctx->GetTempSpaceAllocator(&alloc);
-        if (!status.IsOK()) {
-          return Status(common::ONNXRUNTIME, common::FAIL,
-                        "Memcpy cuda: unable to get an allocator.");
-        }
+      // do we support async copy?
+      auto& src_device = X->Location().device;
+      auto& dst_device = Y->Location().device;
+
+      if (!((src_device.Type() == OrtDevice::CPU && src_device.MemType() != OrtDevice::MemType::CUDA_PINNED) ||
+            (dst_device.Type() == OrtDevice::CPU && dst_device.MemType() != OrtDevice::MemType::CUDA_PINNED))) {
+        auto* gpu_data_transfer = Info().GetDataTransferManager().GetDataTransfer(X->Location().device, Y->Location().device);
+        ORT_RETURN_IF_ERROR(gpu_data_transfer->CopyTensorAsync(*X, *Y, ctx->GetComputeStream()));
+        return Status::OK();
       } else {
-        // If we are copying contents to CPU (op type is "MemcpyToHost"),
-        // the allocator to use to allocate the buffers of the new tensors
-        // in the sequence will be the allocator from the CPU EP
-        auto status = ctx->GetTempSpaceCPUAllocator(&alloc);
-        if (!status.IsOK()) {
-          return Status(common::ONNXRUNTIME, common::FAIL,
-                        "Memcpy cuda: unable to get the CPU allocator.");
-        }
+        auto* gpu_data_transfer = Info().GetDataTransferManager().GetDataTransfer(X->Location().device, Y->Location().device);
+        ORT_RETURN_IF_ERROR(gpu_data_transfer->CopyTensorAsync(*X, *Y, ctx->GetComputeStream()));
+        return Status::OK();
```

Comment:
Can this logic be provided by the DataTransferManager in a reasonably generic way? Otherwise won't we need to replicate the code elsewhere (e.g. the ROCm EP)?

Reply:
ROCM ep already duplicated it quite a bit before this PR :)
The async copy support between Pinned memory and GPU memory is GPU specific, i don't think it is good to have this abstraction in DataTransferManager, right?
 
Weixing used to consider create a "GPU EP" and have Cuda and ROCM as two backends for it. That probably will be a better design. but i don't want to include that in this PR.